---
title: "NICU_nextseq_09.29.22"
author: "Min-Ting"
date: "2022-09-29"
output: html_document
---
#Save the entire list of environments objects
```{r}
# save
save.image(file = "NICU_shotgun.RData")
dir("/Users/min-tinglee/Desktop/Diet_Lipid_Data/nextseq_NICU_20220928")

#load environment
load("/Users/min-tinglee/Desktop/Diet_Lipid_Data/nextseq_NICU_20220928/NICU_shotgun.RData")
```

# libraries
```{r}
library("genefilter")
library("ggpubr")
library("UpSetR")
library("randomcoloR")
library("circlize")
library("RColorBrewer")
library("qiime2R")
library("tidyverse")
library("phyloseq")
library("DESeq2")
library("ggplot2")
library("ComplexHeatmap")
library("curatedMetagenomicData")
library(ggsci)
library(Polychrome)
library(OTUtable)

library(ggforce)

# group higher
library(file2meco)
library(microeco)
library(magrittr)
```

##=========================================================
# Create funciton for converting Metaphlan file to phyloseq
##=========================================================
```{r}
## make function based on
# https://www.biostars.org/p/449688/
metaphlanToPhyloseq <- function(
  tax,
  metadat=NULL,
  simplenames=TRUE,
  roundtointeger=FALSE,
  split="|"){
  ## tax is a matrix or data.frame with the table of taxonomic abundances, rows are taxa, columns are samples
  ## metadat is an optional data.frame of specimen metadata, rows are samples, columns are variables
  ## if simplenames=TRUE, use only the most detailed level of taxa names in the final object
  ## if roundtointeger=TRUE, values will be rounded to the nearest integer
  xnames = rownames(tax)
  shortnames = gsub(paste0(".+\\", split), "", xnames)
  if(simplenames){
    rownames(tax) = shortnames
  }
  if(roundtointeger){
    tax = round(tax * 1e4)
  }
  x2 = strsplit(xnames, split=split, fixed=TRUE)
  taxmat = matrix(NA, ncol=max(sapply(x2, length)), nrow=length(x2))
  colnames(taxmat) = c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species", "Strain")[1:ncol(taxmat)]
  rownames(taxmat) = rownames(tax)
  for (i in 1:nrow(taxmat)){
    taxmat[i, 1:length(x2[[i]])] <- x2[[i]]
  }
  taxmat = gsub("[a-z]__", "", taxmat)
  taxmat = phyloseq::tax_table(taxmat)
  otutab = phyloseq::otu_table(tax, taxa_are_rows=TRUE)
  if(is.null(metadat)){
    res = phyloseq::phyloseq(taxmat, otutab)
  }else{
    res = phyloseq::phyloseq(taxmat, otutab, phyloseq::sample_data(metadat))
  }
  return(res)
}
```

## Import tables for organization
```{r}
# sample sheet from Zeng
meta_zeng <- read_csv("/Users/min-tinglee/Desktop/Diet_Lipid_Data/nextseq_NICU_20220928/NICU_ZengSheet.csv") # C50V7 has two rows

meta_zeng <- meta_zeng[-c(147),] #remove duplicated C50V7

# sample submitted from kevin
meta_kb <- read_csv("/Users/min-tinglee/Desktop/Diet_Lipid_Data/nextseq_NICU_20220928/NICU_Nextera_SampleSubmission.csv") #68 ?


```



##=========================================================
# Metaphlan convert to phyloseq-- Metaphlan4.0
##=========================================================
```{r}
## import Metaphlandata
mpa <- read_csv("/Users/min-tinglee/Desktop/Diet_Lipid_Data/nextseq_NICU_20220928/metaphlan4_merged_abundance_table2.csv")

mpa <- mpa %>%
  column_to_rownames(var = "taxonomy") # move taxonomy to rowname


# remove low quality samples 
mpa <- mpa %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column(var = "SampleID") %>%
  filter(!SampleID %in% c("B21U7", "C61U2", "C70U4", "D48U2", "C65U6")) %>%  # 62 left
  column_to_rownames(var =  "SampleID") %>%
  t()

## Metadata: save only the sequenced ones

meta_new <- mpa %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column(var = "SampleID") %>%
  select(c("SampleID")) %>%
  left_join(meta_zeng, by = "SampleID")


## metadata for phyloseq

row.names(meta_new) <- meta_new$SampleID

sample <- sample_data(meta_new)

## convert mpa to phyloseq
ps = metaphlanToPhyloseq(mpa, metadat = sample)

```


##=========================================================
# Stacked bar- Phylum (with poor quality ones)
##=========================================================
```{r}
## phylum-- ignore NA
ps_phylum_nona <- ps %>%
  tax_glom(taxrank =  "Phylum",
           NArm = TRUE) %>%
  transform_sample_counts(function(x) {x/sum(x)}) %>%
  psmelt()



# how many phylum?
ps_phylum_nona %>%
  group_by(Phylum) %>%
  summarise(n = n()) # 5

#nb.4.col <- colorRampPalette((brewer.pal(8,"Set1")))(5)
## if colorRampPalette(rev(brewer.pal(8,"Set2")))(5) thab reverse 
nb.5.colorblind <- c("#999999", "#56B4E9", "#CC79A7",
          "#F0E442", "#0072B2", "#009E73")



# plot, no NA
ps_phylum_nona %>%
  ggplot(aes(x = SampleID, y = Abundance, fill = Phylum)) +
  geom_bar(stat = "identity",
           position = "fill"
           ) +
  scale_fill_manual(values = nb.5.colorblind) +
  theme_bw() +
  facet_grid( sex ~ antibiotics + birth_mode, scales = "free", space = "free") + #v1
  #facet_grid(antibiotics ~ birth_mode, scales = "free", space = "free") + #v2
  theme(axis.text.x = element_text(angle = 90,
                                   vjust = 1,
                                   hjust = 1),
        text = element_text(size=7))

ggsave(path = "/Users/min-tinglee/Desktop/Diet_Lipid_Data/nextseq_NICU_20220928/Rplot",filename = "stackBar_phylum_nona_facet.pdf", height=8, width=12, device="pdf")

```

##=========================================================
# Stacked bar- Genus
##=========================================================
```{r}
ps_genus_nona <- ps %>%
  tax_glom(taxrank =  "Genus",
           NArm = TRUE) %>%
  transform_sample_counts(function(x) {x/sum(x)}) %>%
  psmelt() %>%
  arrange(Genus)


#temp <- ps_genus_nona[,c(2,3,10)] %>%
  #filter(Genus == "Bacteroides")

# how many genera?
Genus_49 <- ps_genus_nona %>%
  group_by(Genus) %>%
  summarise(n = n()) # 49
  

#nb.18.col <- colorRampPalette((brewer.pal(8,"Set2")))(18)
  #glasbey.colors(32)
nb.49.col <- rev(distinctColorPalette(49))

# assign genus names to color?

names(nb.49.col) <- Genus_49$Genus

# plot
ps_genus_nona %>%
  ggplot(aes(x = SampleID, y = Abundance, fill = Genus)) +
  geom_bar(stat = "identity",
           position = "fill",
           #color = "grey25",
           ) +
  scale_fill_manual(values = nb.49.col) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90,
                                   vjust = 1,
                                   hjust = 1),
        text = element_text(size=7))

ggsave(path = "/Users/min-tinglee/Desktop/Diet_Lipid_Data/nextseq_NICU_20220928/Rplot",filename = "stackBar_Genus_nona.pdf", height=5, width=10, device="pdf")
```


##=========================================================
# Stacked bar- Genus: REMOVE below 0.01
##=========================================================
```{r}
ps_genus_nona_001 <- ps %>%
  tax_glom(taxrank =  "Genus",
           NArm = TRUE) %>%
  transform_sample_counts(function(x) {x/sum(x)}) %>%
  psmelt() %>%
  filter(Abundance > 0.01) %>%
  arrange(Genus)


#temp <- ps_genus_nona[,c(2,3,10)] %>%
  #filter(Genus == "Bacteroides")

# how many genera?
ps_genus_nona_001 %>%
  group_by(Genus) %>%
  summarise(n = n()) # 33

#nb.18.col <- colorRampPalette((brewer.pal(8,"Set2")))(18)
  #glasbey.colors(32)
nb.33.col <- distinctColorPalette(33)

# plot
ps_genus_nona_001 %>%
  ggplot(aes(x = SampleID, y = Abundance, fill = Genus)) +
  geom_bar(stat = "identity",
           position = "fill",
           #color = "grey25",
           ) +
  scale_fill_manual(values = nb.49.col) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90,
                                   vjust = 1,
                                   hjust = 1),
        text = element_text(size=7))

ggsave(path = "/Users/min-tinglee/Desktop/Diet_Lipid_Data/nextseq_NICU_20220928/Rplot",filename = "stackBar_Genus_nona_0.01.pdf", height=5, width=10, device="pdf")



# ********* facet by variable **********

## plot
ps_genus_nona_001 %>%
  ggplot(aes(x = SampleID, y = Abundance, fill = Genus)) +
  geom_bar(stat = "identity",
           position = "fill",
           #color = "grey25",
           ) +
  scale_fill_manual(values = nb.49.col) +
  theme_bw() +
  #facet_grid( sex ~ antibiotics + birth_mode, scales = "free", space = "free") + #v1
  facet_grid(antibiotics ~ birth_mode, scales = "free", space = "free") + #v2
  theme(axis.text.x = element_text(angle = 90,
                                   vjust = 1,
                                   hjust = 1),
        text = element_text(size=7))

ggsave(path = "/Users/min-tinglee/Desktop/Diet_Lipid_Data/nextseq_NICU_20220928/Rplot",filename = "stackBar_Genus_nona_0.01_facet2.pdf", height=8, width=10, device="pdf")
```

##=========================================================
# Stacked bar- Species
##=========================================================
```{r}
ps_species_nona <- ps %>%
  tax_glom(taxrank =  "Species",
           NArm = TRUE) %>%
  transform_sample_counts(function(x) {x/sum(x)}) %>%
  psmelt()

ps_species_nona_001 <- ps %>%
  tax_glom(taxrank =  "Species",
           NArm = TRUE) %>%
  transform_sample_counts(function(x) {x/sum(x)}) %>%
  psmelt() %>%
  filter(Abundance > 0.01) %>%
  arrange(Species)

# how many genera?
ps_species_nona_001 %>%
  group_by(Species) %>%
  summarise(n = n()) # 100 #61 over 1%



#get colors
nb.61.col <- distinctColorPalette(61)
#nb.41.col2 <- colorRampPalette((brewer.pal(8,"Set1")))(41)

# plot
ps_species_nona_001 %>%
  ggplot(aes(x = SampleID, y = Abundance, fill = Species)) +
  geom_bar(stat = "identity",
           position = "fill",
           #color = "grey25",
           ) +
  scale_fill_manual(values = nb.61.col) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90,
                                   vjust = 1,
                                   hjust = 1),
        text = element_text(size=7))

ggsave(path = "/Users/min-tinglee/Desktop/Diet_Lipid_Data/nextseq_NICU_20220928/Rplot",filename = "stackBar_species_nona_001.pdf", height=6, width=14, device="pdf")


## facet ------------------------

ps_species_nona_001 %>%
  ggplot(aes(x = SampleID, y = Abundance, fill = Species)) +
  geom_bar(stat = "identity",
           position = "fill",
           #color = "grey25",
           ) +
  scale_fill_manual(values = nb.61.col) +
  theme_bw() +
  #facet_grid( sex ~ antibiotics + birth_mode, scales = "free", space = "free") + #v1
  facet_grid(antibiotics ~ birth_mode, scales = "free", space = "free") + #v2
  theme(axis.text.x = element_text(angle = 90,
                                   vjust = 1,
                                   hjust = 1),
        text = element_text(size=7))

ggsave(path = "/Users/min-tinglee/Desktop/Diet_Lipid_Data/nextseq_NICU_20220928/Rplot",filename = "stackBar_species_nona_001_facet.pdf", height=8, width=14, device="pdf")


```
##=========================================================
# Sort 
##=========================================================
```{r}
### *************** Genus ***************

ps_genus_nona_001 %>%
  mutate(genus_ordering = fct_reorder(Genus, Abundance)) %>%
  ggplot(aes(x = genus_ordering, y = Abundance)) +
  geom_boxplot(aes(fill = Genus, alpha = 0.5)) +
  geom_jitter(aes(color = Genus)) +
  scale_color_manual(values = nb.33.col) +
  scale_fill_manual(values = nb.33.col) +
  theme(axis.text.x=element_text(angle = 90,
                                   vjust = 1,
                                   hjust = 1, 
                                   size = 7))

ggsave(path = "/Users/min-tinglee/Desktop/Diet_Lipid_Data/nextseq_NICU_20220928/Rplot",filename = "genus_sort.pdf", height=6, width=12, device="pdf")

### *************** Genus, facet ***************
ps_genus_nona_001 %>%
  mutate(genus_ordering = fct_reorder(Genus, Abundance)) %>%
  ggplot(aes(x = genus_ordering, y = Abundance)) +
  geom_boxplot(aes(fill = Genus, alpha = 0.5)) +
  geom_jitter(aes(color = Genus)) +
  scale_color_manual(values = nb.33.col) +
  scale_fill_manual(values = nb.33.col) +
  facet_grid(antibiotics ~ birth_mode, scales = "free", space = "free") +
  theme(axis.text.x=element_text(angle = 90,
                                   vjust = 1,
                                   hjust = 1, 
                                   size = 7))

ggsave(path = "/Users/min-tinglee/Desktop/Diet_Lipid_Data/nextseq_NICU_20220928/Rplot",filename = "genus_sort_facet.pdf", height=8, width=12, device="pdf")
```
##=========================================================
# Pathway 
##=========================================================

# ***********
# import data and scale
# ***********
```{r}
## *******
# This is the unstratified one 
## *******

path_unstratified <- read_csv("/Users/min-tinglee/Desktop/Diet_Lipid_Data/nextseq_NICU_20220928/humann3.5_merged_NICU_pathabundance-cpm_unstratified.csv")


# remove those with over half 0

path_unstratified.simple <- path_unstratified[rowSums(path_unstratified == 0) <= 34, ] #from 455 to 283

## ***** scale ******

path_unstratified.simple.mx <- path_unstratified.simple %>%
  filter(!Pathway == "UNMAPPED") %>%
  filter(!Pathway == "UNINTEGRATED") %>% # 281 left
  pivot_longer(-Pathway, names_to = "SampleID", values_to = "CPM") %>%
  group_by(SampleID, Pathway) %>%
  summarise(avg_CPM = mean(CPM)) %>%
  filter(!SampleID == "B21U7", !SampleID =="C61U2", !SampleID =="C70U4",
         !SampleID =="D48U2", !SampleID =="C65U6") %>%
  pivot_wider(names_from = "SampleID", values_from = "avg_CPM") %>%
  column_to_rownames(var = "Pathway") %>%
  t() %>%
  as.matrix() %>%
  scale() %>%
  t()
```

## **************
# group pathway to higher Metacyc
## **************
```{r}
# use origianl pathway table

path_ori_meco <- humann2meco(abund_table = "/Users/min-tinglee/Desktop/Diet_Lipid_Data/nextseq_NICU_20220928/humann3.5_merged_NICU_pathabundance-cpm_stratified_filter.tsv" , db = "MetaCyc")

# rel = FALSE do not use relative abundance
path_ori_meco$cal_abund(select_cols = 1:3, rel = FALSE)
path_ori_meco$taxa_abund$Superclass2 %<>% .[!grepl("unclass", rownames(.)), ]

path_ori_meco1 <- trans_abund$new(path_ori_meco, taxrank = "Superclass2", ntaxa = 10)

## ******** table
as.data.frame(path_ori_meco$tax_table) %>%
  rownames_to_column(var = "Pathway") %>%
  write.csv("/Users/min-tinglee/Desktop/Diet_Lipid_Data/nextseq_NICU_20220928/Rplot/path_groupMeco2.csv")
## *********

pdf(file ="/Users/min-tinglee/Desktop/Diet_Lipid_Data/nextseq_NICU_20220928/Rplot/path_groupMeco_top10.pdf", height=10, width=12)

path_ori_meco1$plot_bar(ylab_title = "Abundance (CPM)") +
  coord_flip() +
  theme_bw()

dev.off()

##
path_ori_meco$tax_table

as.data.frame(path_ori_meco1$data_abund) %>%
  group_by(Taxonomy) %>%
  summarise(n = n()) ## 41


## 
nb.41.col <- distinctColorPalette(41)

pdf(file ="/Users/min-tinglee/Desktop/Diet_Lipid_Data/nextseq_NICU_20220928/Rplot/path_groupMeco.pdf", height=6, width=25)

as.data.frame(path_ori_meco1$data_abund) %>%
  left_join(meta_new, by = "SampleID") %>%
  ggplot(aes(x = Sample, y = Abundance,
             color = Taxonomy, group = Taxonomy)) +
  geom_point(size = 1) +
  scale_color_manual(values = nb.41.col) +
  geom_line()+
  #scale_y_discrete(expand=c(0.3, 0))+
  xlab('SampleID') +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90,
                                   vjust = 1,
                                   hjust = 1,
                                   size = 5),
        text = element_text(size=10))
  #facet_grid(antibiotics ~ birth_mode, scales = "free", space = "free")
  

dev.off()



###
pdf(file ="/Users/min-tinglee/Desktop/Diet_Lipid_Data/nextseq_NICU_20220928/Rplot/path_groupMeco_facet.bar.pdf", height=12, width=25)

as.data.frame(path_ori_meco1$data_abund) %>%
  left_join(meta_new, by = "SampleID") %>%
  ggplot(aes(x = Sample, y = Abundance,
             fill = Taxonomy)) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values = nb.41.col) +
  xlab('SampleID') +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90,
                                   vjust = 1,
                                   hjust = 1,
                                   size = 5),
        text = element_text(size=10)) +
  facet_grid(antibiotics ~ birth_mode, 
             scales = "free", space = "free") + # v2 no facet
  coord_flip()

dev.off()

```


# ******* 
# Heatmap
# ******* 
```{r}

## ***** Column annotation ******

path_colAnn <- path_unstratified.simple.mx %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column(var = "SampleID") %>%
  select(c("SampleID")) %>%
  left_join(meta_new)

path_colAnn$Feeding_Notes <- path_colAnn$Feeding_Notes %>% replace_na("No_info")

path_colAnn <- path_colAnn[-c(5)] # feeding notes too messy

path_colAnn <- path_colAnn %>%
  column_to_rownames(var = "SampleID")


sex_color <- c("F" = "#e04c41", "M" = "#236cd6", "No_info" = "#2e3e51")
antibiotic_color <- c("no" = "#9659b2", "yes" = "#f2b94a", "No_info" = "#2e3e51")
deliver_color <- c("C" = "#3db182", "V" = "#ffcccd","No_info" = "#2e3e51")


## ***** Row annotation ******
path_super_clean <- read_csv("/Users/min-tinglee/Desktop/Diet_Lipid_Data/nextseq_NICU_20220928/Rplot/path_groupMeco2.csv")

path_super_clean2<- path_super_clean[,c(2:5)] %>%
  unique()


path_rowAnn <- path_unstratified.simple.mx %>%
  as.data.frame() %>%
  rownames_to_column(var = "Pathway") %>%
  left_join(path_super_clean2, by = "Pathway") %>%
  select(c("Pathway", "Superclass2"))

path_rowAnn <- path_rowAnn %>%
  as.data.frame() %>%
  remove_rownames() %>%
  column_to_rownames(var = "Pathway")

path_rowAnn  %>%
  group_by(Superclass2) %>%
  summarise(n = n())  # 36

nb.36.col <- distinctColorPalette(36)
names(nb.36.col) <- unique(path_rowAnn$Superclass2)


# annotation list

path_color <- list(sex = sex_color,
                   birth_mode = deliver_color,
                   antibiotics = antibiotic_color,
                   Superclass2 = nb.36.col) ## make it a list


## ***** PLOT ******

pdf(file ="/Users/min-tinglee/Desktop/Diet_Lipid_Data/nextseq_NICU_20220928/Rplot/path_unstratified_anno.pdf", height=10, width=12)

path_unstratified.simple.mx %>%
  pheatmap(cluster_cols = T,
           #color=colorRampPalette(c("navy", "white", "red"))(100),
           annotation_col = path_colAnn, #change here
           annotation_row = path_rowAnn,
           annotation_colors = path_color,
           fontsize_row = 2,
           fontsize_col = 5)

dev.off()
```





